diff --git a/DMCompiler/DM/DMProc.cs b/DMCompiler/DM/DMProc.cs
index bf8e5ad..5db364a 100644
--- a/DMCompiler/DM/DMProc.cs
+++ b/DMCompiler/DM/DMProc.cs
@@ -227,7 +227,8 @@ internal sealed class DMProc {
 
                 arguments.Add(new ProcArgumentJson {
                     Name = parameter.Name,
-                    Type = argumentType.Type
+                    Type = argumentType.Type,
+                    TypePath = parameter.Type?.ToString()
                 });
             }
         }
diff --git a/DMCompiler/Json/DreamProcJson.cs b/DMCompiler/Json/DreamProcJson.cs
index eb9f4a0..af08312 100644
--- a/DMCompiler/Json/DreamProcJson.cs
+++ b/DMCompiler/Json/DreamProcJson.cs
@@ -26,6 +26,7 @@ public sealed class ProcDefinitionJson {
 public struct ProcArgumentJson {
     public required string Name { get; init; }
     public DMValueType Type { get; init; }
+    public string? TypePath { get; init; }
 }
 
 public struct LocalVariableJson {
